### PR TITLE
Updated Sobol' indices from PCE

### PR DESCRIPTION
**Pull request by mbaudin47**
The goal of this PR is to improve the doc and improve the accuracy of implementation of the `FunctionalChaosSobolIndices` class (see [the current doc](http://openturns.github.io/openturns/latest/user_manual/response_surface/_generated/openturns.FunctionalChaosSobolIndices.html)).

### The `FunctionalChaosSobolIndices` API help page

The first topic is the API help page of `FunctionalChaosSobolIndices`. A consistent doc of `FunctionalChaosSobolIndices` should rely on the theory page of Sobols' indices, [in the theory help](http://openturns.github.io/openturns/latest/theory/reliability_sensitivity/sensitivity_sobol.html). But that page does not present the Sobol' indices: it presents the estimators of Sobol' indices using sampling methods. One of the issues of the API help page of `FunctionalChaosSobolIndices` is that the doc does not state clearly what exactly is estimated by each method. The next table presents the map from each of the 6 possible Sobol' indices to the corresponding methods.

| Sobol' index | Method |
|---|---|
| First order (simple) | `getSobolIndex(i)` |
| Total (simple) | `getSobolTotalIndex(i)` |
| Group interaction  | `getSobolIndex(group)` |
| Group total interaction | `getSobolTotalIndex(group)` |
| Group first order (closed) | `getSobolGroupedIndex(group)` |
| Group total | `getSobolGroupedTotalIndex(group)` |

**Table 1.** The mapping from Sobol' indices to the methods of `FunctionalChaosSobolIndices` .

In the current doc of `FunctionalChaosSobolIndices`, there is no accurate definition of the purpose of each method. This is confusing, since the cases where the argument is a single integer or a sequence of integers are linked in a non-obvious way: the single integer case is a special example of the more general formula.

image

**Figure 1.** The doc of `getSobolGroupedIndex` does not present neither the definition of the Sobol' index, nor on the equations to compute the Sobol' index from the PCE.

### The `Sensitivity analysis using Sobol indices` theory help page

The [theory help page of Sobol' indices](http://openturns.github.io/openturns/latest/theory/reliability_sensitivity/sensitivity_sobol.html) does not present Sobol' indices, but estimators of it. Another related page is [SobolIndicesAlgorithm](http://openturns.github.io/openturns/latest/user_manual/_generated/openturns.SobolIndicesAlgorithm.html) which presents sampling based estimators of Sobol' indices. But part of this API help page should be presented in the theory help page. In other words, the theory help page does not present the theory of Sobol indices and some part of the API help page should be presented in the theory help page. 

Furthermore, the presentation of the estimators of the Sobol' indices is obscure and does not clearly introduces the pick-freeze estimator.

### The `SobolIndicesAlgorithm` API help page

The API help page [SobolIndicesAlgorithm](http://openturns.github.io/openturns/latest/user_manual/_generated/openturns.SobolIndicesAlgorithm.html) introduces Sobol' indices. This should be done in the [theory help page of Sobol' indices](http://openturns.github.io/openturns/latest/theory/reliability_sensitivity/sensitivity_sobol.html).

The notation of a vector are either wrong or obscure:

image

**Figure 2.** The notation of a vector in `SobolIndicesAlgorithm` API help page.

This may mean that the vector X is raised to the power 1, power 2, etc. up to the power nX. The exponent is actually an index.

The equations have no name and are wrong:

image

**Figure 3.** A wrong version of Möbius formula in `SobolIndicesAlgorithm` API help page.

This is the wrong version of the Möbius inversion formula. The correct one is:

image

**Figure 4.** The correct version of Möbius formula.

The notations of total variance and total Sobol' index are unclassical:

image

**Figure 5.** The notations for total variance and total Sobo'l index in `SobolIndicesAlgorithm` API help page.

The classical way to denote a total Sobol' index with by using T as exponent.

image

**Figure 6.** The notation used in (Da Veiga, et al., 2021) page 53.

This is still a questionable notation. First, the upper case letter "S" may mean a random variable, which is not. Then the exponent upper case letter "T" may mean "transpose" in linear algebra. Then "S^T" may mean the matrix S transposed. This is, however, the classical notation, so that I suggest to stick with it.

### Implementation of FunctionalChaosSobolIndices

The current implementation of `getSobolGroupedTotalIndex()` method is based on the complementary indices:

https://github.com/openturns/openturns/blob/f2538ea192405ec2360edb379955c21eff66a2ad/lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx#L352

This can lead to an accuracy issue when the index is close to 0. In that case, the `complementaryFirstIndex` is close to 1 and the subtraction with 1 can lead to a [catastrophic cancellation](https://en.wikipedia.org/wiki/Catastrophic_cancellation).

The implementation of `getSobolIndex()` is efficient, but rather obscure:

https://github.com/openturns/openturns/blob/f2538ea192405ec2360edb379955c21eff66a2ad/lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx#L203

The trick used here:

https://github.com/openturns/openturns/blob/f2538ea192405ec2360edb379955c21eff66a2ad/lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx#L215

is clever, but hides the actual meaning of the selection procedure. This is because the implementation is based on `std::max_element` of an `Indices` since the methods that we need are not implemented by `Indices`, i.e. the infinite norm. Furthermore, the implementation of `Indices` provides the `contains()` method, which is not used here although it is relevant. The `getSobolTotalIndex()` and `getSobolGroupedIndex()` methods have the same flaws.

## Summary of the changes

All in all, this PR includes the following improvements.
- Theory help pages.
  - Fill the "Sensitivity analysis using Sobol indices" theory help page. Move the theory content of SobolIndicesAlgorithm into this theory help page to avoid duplicated content.
  - Create a new theory help page to present the methods of PCE-based Sobol' indices. Document the _part of variance_ index which is printed by the `print` function of `FunctionalChaosSobolIndices` (former summary() method), but undocumented.
- API help pages.
  - Add link from the API help pages to the theory help pages. 
  - Fill the API help page of `FunctionalChaosSobolIndices` with a consistent documentation of the PCE-based estimators. 
  - Update API help page of `SobolIndicesAlgorithm`. Move the theory content into the corresponding help page. Fix notation issues.
- C++ source code.
  - Created new `Indices.normInf()` and `Indices.norm1()` to compute the corresponding norms.
  - Updated the implementation of the `FunctionalChaosSobolIndices.getSobolGroupedTotalIndex()` method in order to avoid the loss of accuracy of the current implementation.
  - Updated the implementation  of the`FunctionalChaosSobolIndices.getSobolIndex()` method to clarify the code and use appropriate methods of the `Indices` class. Same for `getSobolTotalIndex()`.
- Unit tests.
  - Fill the `FunctionalChaosSobolIndices` unit tests with in-depth Ishigami and G-Sobol' testing at C++ and Python levels.

- Created new methods to the `Indices` class: `normInfinite()` and `normOne()`. Use these methods in the `FunctionalChaosSobolIndices` to clarify the considered indices. 

**TODO-List**
- [X] Remove the duplicate `getSobolGroupedInteractionIndex()` of `getSobolIndex()`. 
- ~~[ ] In `contains()`, use `str::find`.~~ This will be done by @jschueller at the `Collection` level, because `Indices` is a collection.
- [X] Rename `Indices.normInfinite()` into `Indices.normInf()`.
- [X] Rename `Indices.normOne` into `Indices.norm1()`.

### CODE REVIEW

File: lib/src/Base/Type/Indices.cxx [hunk 1]
```diff
@@ -80,4 +80,38 @@ Indices Indices::complement(const UnsignedInteger n) const
...
+  UnsignedInteger norm = 0;
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    if (operator[](i) > norm) norm = operator[](i);
+  }
+  return norm;
+}
+
+/* Returns the 1-norm of the indices */
+UnsignedInteger Indices::normOne() const
```
Comment (regislebrun) [inline lib/src/Base/Type/Indices.cxx:106]: See previous comment

File: lib/src/Base/Type/Indices.cxx [hunk 2]
```diff
@@ -80,4 +80,38 @@ Indices Indices::complement(const UnsignedInteger n) const
...
+  if (!(size > 0)) return false;
+  for (UnsignedInteger i = 0; i < size; ++i)
+    if (operator[](i) == n) return true;
+  return false;
+}
+
+/* Returns the infinite-norm of the indices */
+UnsignedInteger Indices::normInfinite() const
+{
+  const UnsignedInteger size = getSize();
```
Comment (regislebrun) [inline lib/src/Base/Type/Indices.cxx:96]: By the way, use the implementation based on std::max_element, much more efficient and clearer to read (it is the code you removed in `getSobolGroupedIndex`

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx [hunk 3]
```diff
@@ -294,63 +294,168 @@ Scalar FunctionalChaosSobolIndices::getSobolTotalIndex(const UnsignedInteger var
...
+  // Sum the contributions of all the coefficients associated to a basis vector involving only the needed variables
+  Scalar totalVariance = 0.0;
+  const UnsignedInteger groupDimension = variablesGroup.getSize();
+  bool mustInclude;
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    if (coefficientIndices[i] > 0)
+    {
+      const Scalar coefficientI = coefficients(i, 0);
+      Indices multiIndices(enumerateFunction(coefficientIndices[i]));
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx:371]: const

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx [hunk 4]
```diff
@@ -294,63 +294,168 @@ Scalar FunctionalChaosSobolIndices::getSobolTotalIndex(const UnsignedInteger var
...
+  const EnumerateFunction enumerateFunction(functionalChaosResult_.getOrthogonalBasis().getEnumerateFunction());
+  // Sum the contributions of all the coefficients associated to a basis vector involving only the needed variables
+  Scalar totalVariance = 0.0;
+  bool mustInclude;
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    if (coefficientIndices[i] > 0)
+    {
+      const Scalar coefficientI = coefficients(i, 0);
+      Indices multiIndices(enumerateFunction(coefficientIndices[i]));
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx:420]: const

File: lib/src/Base/Type/openturns/Indices.hxx [hunk 5]
```diff
@@ -88,6 +88,15 @@ public:
   /** Compute the complement between the indices and {0,...,n-1} for a given n */
   Indices complement(const UnsignedInteger n) const;
 
+  /** Check if the integer is in the indices */
+  Bool contains(const UnsignedInteger n) const;
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Base/Type/openturns/Indices.hxx:92]: @jschueller Any reason why `Collection::contains` is not implemented using `std::find`? I would have expected something like:
```
  inline
  Bool contains(T val) const
  {
    return (std::find(begin(), end(), val) != end());
  }
```

[the remark this answers is not in the document]
Comment (jschueller) [reply lib/src/Base/Type/openturns/Indices.hxx:92]: Let's try this in the misc2 branch

File: lib/src/Base/Type/Indices.cxx [hunk 6]
```diff
@@ -80,4 +80,38 @@ Indices Indices::complement(const UnsignedInteger n) const
...
+  if (!(size > 0)) return false;
+  for (UnsignedInteger i = 0; i < size; ++i)
+    if (operator[](i) == n) return true;
+  return false;
+}
+
+/* Returns the infinite-norm of the indices */
+UnsignedInteger Indices::normInfinite() const
```
[the remark this answers is not in the document]
Comment (mbaudin47) [reply lib/src/Base/Type/Indices.cxx]: Ok. Fixed at e6509ef3f1aa53a28a90679b27eb14465b1cc0ef

File: lib/src/Base/Type/Indices.cxx (hunk 2 above)
Comment (mbaudin47) [reply lib/src/Base/Type/Indices.cxx]: Ok. Fixed at 560c7cff65dae314219b203a994db37cad01d924

File: lib/src/Base/Type/Indices.cxx (hunk 1 above)
Comment (mbaudin47) [reply lib/src/Base/Type/Indices.cxx]: Ok. Fixed at e6509ef3f1aa53a28a90679b27eb14465b1cc0ef

File: lib/src/Base/Type/openturns/Indices.hxx (hunk 5 above)
[the remark this answers is not in the document]
Comment (mbaudin47) [reply lib/src/Base/Type/openturns/Indices.hxx:92]: This is done by @jschueller  at the Collection level in https://github.com/jschueller/openturns/commit/17535db7bccf5c88ea37cf3ba4b83184c0c4c565. This is why I consider that this issue is resolved in the current PR.

File: lib/src/Base/Type/Indices.cxx [hunk 7]
```diff
@@ -80,4 +80,14 @@ Indices Indices::complement(const UnsignedInteger n) const
   return result;
 }
 
+/* Return true if the integer n in is the indices */
+Bool Indices::contains(const UnsignedInteger n) const
+{
+  const UnsignedInteger size = getSize();
```
[the remark this answers is not in the document]
Comment (mbaudin47) [reply lib/src/Base/Type/Indices.cxx:86]: See https://github.com/jschueller/openturns/commit/17535db7bccf5c88ea37cf3ba4b83184c0c4c565 for a proper fix in the `Collection` class.

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx (hunk 3 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx:381]: Ok. I fixed the similar const issue at e008e7f54f23bedd45db149731931b48c32e5608.

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx (hunk 4 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx]: Ok. I fixed the similar const issue at e008e7f54f23bedd45db149731931b48c32e5608.

**Comment by mbaudin47** [comment]
@vchabri : I would highly appreciate your review!

File: lib/src/Base/Type/Indices.cxx [hunk 8]
```diff
@@ -80,4 +80,23 @@ Indices Indices::complement(const UnsignedInteger n) const
...
+}
+
+/* Returns the 1-norm of the indices */
+UnsignedInteger Indices::norm1() const
+{
+  const UnsignedInteger size = getSize();
+  UnsignedInteger norm = 0;
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    norm += operator[](i);
```
Comment (jschueller) [inline lib/src/Base/Type/Indices.cxx:97]: better use std::accumulate

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx [hunk 9]
```diff
@@ -294,65 +304,109 @@ Scalar FunctionalChaosSobolIndices::getSobolTotalIndex(const UnsignedInteger var
...
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    if (coefficientIndices[i] > 0)
+    {
+      const Scalar coefficientI = coefficients(i, 0);
+      const Indices multiIndices(enumerateFunction(coefficientIndices[i]));
+      // Compute denominator
+      if (multiIndices.normInf() > 0) totalVariance += coefficientI * coefficientI;
+      // Compute numerator
+      if (multiIndices.normInf() == 0)
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx:385]: we could try to compute normInf once

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx [hunk 10]
```diff
@@ -294,65 +304,109 @@ Scalar FunctionalChaosSobolIndices::getSobolTotalIndex(const UnsignedInteger var
...
+  if (!variablesGroup.check(inputDimension)) throw InvalidArgumentException(HERE) << "The variable indices of a Sobol' index must be in the range [0, dim-1] and must be different.";
+  const Sample coefficients(functionalChaosResult_.getCoefficients().getMarginal(marginalIndex));
+  const Indices coefficientIndices(functionalChaosResult_.getIndices());
+  const UnsignedInteger size = coefficients.getSize();
+  Scalar covarianceVariables = 0.0;
+  const EnumerateFunction enumerateFunction(functionalChaosResult_.getOrthogonalBasis().getEnumerateFunction());
+  // Sum the contributions of all the coefficients associated to a basis vector involving only the needed variables
+  Scalar totalVariance = 0.0;
+  const UnsignedInteger groupDimension = variablesGroup.getSize();
+  bool mustInclude;
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx:375]: we could initialize it once to false

File: python/src/Indices_doc.i.in [hunk 11]
```diff
@@ -65,6 +76,7 @@ Parameters
 ----------
 bound : positive int
     The value of :math:`n`.
+    Why does Sphinx hide this line ????
```
Comment (jschueller) [inline python/src/Indices_doc.i.in:79]: ?

**Comment by vchabri** [comment]
From what I read (your "What should be fixed" and, mostly, the corresponding modifications in the doc together with some parts of the code, but not everything -- to be honest!), I fully agree with your remarks and proposed changes. Congrats.

**Review by regislebrun — APPROVED** [review]

File: lib/src/Base/Type/Indices.cxx (hunk 8 above)
Comment (mbaudin47) [reply lib/src/Base/Type/Indices.cxx:97]: Ok. Fixed at 90332ee59411467c21b7a6ca3c43eaff4c015764

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx (hunk 9 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx:386]: Ok. Fixed at 3f30713e3d3d5f5de78d890494137f0632a96ab9

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx (hunk 10 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx]: Ok. Fixed at e3cd3cfe3cb97c40515f5fbf3de21a1d2300b3a9

File: python/src/Indices_doc.i.in (hunk 11 above)
Comment (mbaudin47) [reply python/src/Indices_doc.i.in:79]: I think I was wrong:
image

Comment (mbaudin47) [reply python/src/Indices_doc.i.in]: Ok. Fixed at e75ce8cf082ba2204507ed9543968b8eca366212

Comment (mbaudin47) [reply python/src/Indices_doc.i.in]: This is fixed at : 
image

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx [hunk 12]
```diff
@@ -294,65 +305,111 @@ Scalar FunctionalChaosSobolIndices::getSobolTotalIndex(const UnsignedInteger var
...
-          // Set the exponents corresponding to the group to zero
-          for (UnsignedInteger j = 0; j < orderSobolIndice; ++j)
-            multiIndices[variableIndices[j]] = 0;
-          // Now check that all the indices are zero
-          if (*std::max_element(multiIndices.begin(), multiIndices.end()) == 0)
-            covarianceVariables += coefficientI * coefficientI;
-        } // *std::max_element(multiIndices.begin(), multiIndices.end()) > 0
-      } // if coefficientI <> 0
+          // Check that each variable having a nonzero degree is in the group
+          if (multiIndices[variableIndex] > 0 and not variablesGroup.contains(variableIndex))
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx:349]: "and", "not" are not allowed in c++

File: python/doc/theory/reliability_sensitivity/sensitivity_sobol_from_pce.rst [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+.. topic:: Examples:
+
+    - See :doc:`/auto_reliability_sensitivity/sensitivity_analysis/plot_sensitivity_sobol`
+    - See :doc:`/auto_reliability_sensitivity/sensitivity_analysis/plot_functional_chaos_sensitivity`
+
+.. topic:: References:
+
+    - Sudret, B. (2006). Global sensitivity analysis using polynomial chaos expansions. In. Proceedings of the 5th International Conference on Computational Stochastic Mechanics (CSM5), Rhodos (2006)
+    - Sudret, B. (2008). Global sensitivity analysis using polynomial chaos expansions. *Reliability engineering & system safety*, *93* (7), 964-979.
+    - Knio, O. M., & Le Maitre, O. P. (2006). Uncertainty propagation in CFD using polynomial chaos decomposition. *Fluid dynamics research*, *38* (9), 616.
```
Comment (jschueller) [inline python/doc/theory/reliability_sensitivity/sensitivity_sobol_from_pce.rst:251]: better move this to the bibliography as its cited in the other doc too

**Review by jschueller — APPROVED** [review]
did not find any mistake in the doc
so I have to approve :[ pending minor fixes
good job

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx [hunk 14]
```diff
@@ -294,65 +305,111 @@ Scalar FunctionalChaosSobolIndices::getSobolTotalIndex(const UnsignedInteger var
 }
 
 /* Sobol grouped (first order) index accessor */
-Scalar FunctionalChaosSobolIndices::getSobolGroupedIndex(const Indices & variableIndices,
+Scalar FunctionalChaosSobolIndices::getSobolGroupedIndex(const Indices & variablesGroup,
     const UnsignedInteger marginalIndex) const
 {
+  Log::Show(Log::ALL);
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx:311]: I think that this is an intermediate debug code.

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx (hunk 12 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx:349]: Ok. Fixed.

File: python/doc/theory/reliability_sensitivity/sensitivity_sobol_from_pce.rst (hunk 13 above)
Comment (mbaudin47) [reply python/doc/theory/reliability_sensitivity/sensitivity_sobol_from_pce.rst]: Ok. Fixed at 3af5ad46b

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx (hunk 14 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx]: Fixed at 3408f3ab4

**Comment by jschueller** [comment]
@sofianehaddad good to go ?

**Review by sofianehaddad — APPROVED** [review]

**Pull request merged**